odoo/addons/base/models/res_config.py

```diff
@@ -451,12 +447,18 @@ def _get_classified_fields(self):
                     'other':   ['other_field', ...],
                 }
         """
-        IrModule = self.env['ir.module.module']
+        IrModule = self.env['ir.module.module'].sudo()
+        IrModelData = self.env['ir.model.data']
         Groups = self.env['res.groups']
-        ref = self.env.ref
+
+        def ref(xml_id):
+            model, id = IrModelData.xmlid_to_res_model_res_id(xml_id)
+            return self.env[model].browse(id)
```

Comment:
Why not just `self.env.ref`?

Reply:
`ref` triggers one exists call every time, which triggers one SQL query for "nothing"
api.ref --> ir.model.xmlid_to_object --> models.exists() --> SELECT id from * WHERE id in (ids)

Reply:
Technically the fields groups might be gone, but I see the point. Still, this should be made more explicit.

But maybe it'd be worth resolving the groups in bulk upfront instead of one-by-one? Even if xmlid_lookup is cached, by collecting all the groups we need to check first we could perform just one query simultanously resolving the xids and checking that the groups are still present in the system.

Reply:
Well, the fact is that it crashes actually (due to the ref + exists) and will still crash after...

If the record was cleanly deleted through the ORM, you will get a clean ValueError like before (because the cache is emptied when an ir.model.data is deleted, which happens if you delete a group with xml_id).
If the record was dirtily deleted (ondelete=cascade / manually deleted in sql) and/or the xml_id was deleted in SQL, then you will have an error (less clear than before with the exists) when reading any field on the group.
I don't believe it's a big loss/change.

As the cache is cleared every time an ir.model.data is unlinked (which would happen if you deleted the group linked to it), I'll still raise a ValueError if the record doesn't exist

Reply:
> Well, the fact is that it crashes actually (due to the ref + exists) and will still crash after...

Crashing at some later point because of corrupted data is a pain in the ass to debug.

Reply:
It will crash at the same moment, unless the user modified a record (res.group/ir.model.data) directly in the database.
```python
self.env.ref('wrong_record')
raise ValueError('No record found for unique ID %s. It may have been deleted.' % (xmlid))

# After my changes
self.env['ir.model.data'].xmlid_to_res_model_res_id('wrong_xmlid')
raise ValueError('External ID not found in the system: %s' % xmlid)

# Side behavior, if there is a cached value which may not have been cleared because the operation
# was done directly in SQL, not through the ORM, and thus the cached data still exists.
self.env['ir.model.data'].xmlid_to_res_model_res_id('wrong_xmlid') # id in cache returned
model.browse(id).field # ORM error because invalid id, the record was deleted in database
# odoo.exceptions.MissingError: Record does not exist or has been deleted
```

Reply:
> It will crash at the same moment, unless the user deleted the record (res.group) directly in the database.

Or they have a long-migrated database which has data inconsistencies, pretty sure we've had cases where the xids were not properly cleaned up alongside their record in the past.

Reply:
Can't we consider that for res.groups referenced in settings, the state of any db is valid ?
If not, they wouldn't be able to open their settings, which is a bug quite big and easy to spot...